File: src/main/java/io/protobase/f7/models/CellQuery.java
package io.protobase.f7.models;

import com.google.common.base.MoreObjects;
import com.google.common.collect.Range;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.utils.AlphaUtils;

import java.util.Objects;
import java.util.Optional;

/**
 * A range query represents a query of one or more cells over a single grid. It is composed of two dimensional ranges:
 * rows and columns, both of which can be unbounded at the lower, or upper ends of the range.
 */
public class CellQuery extends BaseObject {
  /**
   * Name of the grid being queried.
   */
  private String grid;
  /**
   * Range of columns.
   */
  private Range<Integer> columns;
  /**
   * Range of rows.
   */
  private Range<Integer> rows;

  public CellQuery(String grid, Range<Integer> columns, Range<Integer> rows) {
    this.grid = grid;
    this.columns = columns;
    this.rows = rows;
  }

  public static Builder builder() {
    return new Builder();
  }

  public static Builder builder(CellQuery query) {
    return new Builder(query);
  }

  public Optional<String> getGrid() {
    return Optional.ofNullable(grid);
  }

  public Integer getStartColumn() {
    return columns.lowerEndpoint();
  }

  public Integer getEndColumn() {
    return columns.upperEndpoint();
  }

  public Integer getStartRow() {
    return rows.lowerEndpoint();
  }

  public Integer getEndRow() {
    return rows.upperEndpoint();
  }

  /**
   * Does this query intersect with the other one?
   *
   * @param other - other query.
   * @return true if at least one cell overlaps.
   */
  public boolean intersects(CellQuery other) {
    return other.columns.isConnected(this.columns) && other.rows.isConnected(this.rows);
  }

  /**
   * Bound this query using an upper column and upper row. Will default to the lower of the upper column passed in,
   * and the upper column bound existing on the columns range. Same for rows.
   *
   * @param upperColumn - upper endpoint for columns to default to.
   * @param upperRow    - upper endpoint of rows to default to.
   * @return bounded query.
   */
  public CellQuery toBounded(int upperColumn, int upperRow) {
    if (!Range.closed(0, upperColumn).isConnected(columns) && !Range.closed(0, upperRow).isConnected(rows)) {
      return this;
    }
    return CellQuery.builder(this)
        .columnsBetween(this.columns.lowerEndpoint(),
            Math.min(upperColumn, columns.hasUpperBound() ? columns.upperEndpoint() : upperColumn))
        .rowsBetween(this.rows.lowerEndpoint(),
            Math.min(upperRow, rows.hasUpperBound() ? rows.upperEndpoint() : upperRow))
        .build();
  }

  @Override
  public Object[] significantAttributes() {
    return new Object[]{
        grid,
        columns,
        rows
    };
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("grid", grid)
        .add("columns", columns)
        .add("rows", rows)
        .toString();
  }

  public static class Builder {
    private String gridName;
    private Range<Integer> columns;
    private Range<Integer> rows;

    public Builder() {
    }

    public Builder(CellQuery query) {
      this.gridName = query.grid;
      this.columns = query.columns;
      this.rows = query.rows;
    }

    public Builder grid(String gridName) {
      this.gridName = gridName;
      return this;
    }

    public Builder columnsBetween(String low, String high) {
      columns = Range.closed(AlphaUtils.columnToInt(low), AlphaUtils.columnToInt(high));
      return this;
    }

    public Builder columnsBetween(int low, int high) {
      columns = Range.closed(low, high);
      return this;
    }

    public Builder rowsBetween(String low, String high) {
      rows = Range.closed(Integer.parseInt(low) - 1, Integer.parseInt(high) - 1);
      return this;
    }

    public Builder rowsBetween(int low, int high) {
      rows = Range.closed(low, high);
      return this;
    }

    public Builder openColumnsStartingAt(String startColumn) {
      columns = Range.atLeast(AlphaUtils.columnToInt(startColumn));
      return this;
    }

    public Builder openColumnsStartingAt(int startColumn) {
      columns = Range.atLeast(startColumn);
      return this;
    }

    public Builder openColumnsStartingAtZero() {
      columns = Range.atLeast(0);
      return this;
    }

    public Builder openRowsStartingAt(String startRow) {
      rows = Range.atLeast(Integer.parseInt(startRow) - 1);
      return this;
    }

    public Builder openRowsStartingAt(int startRow) {
      rows = Range.atLeast(startRow);
      return this;
    }

    public Builder openRowsStartingAtZero() {
      rows = Range.atLeast(0);
      return this;
    }

    public Builder expand(CellQuery query) {
      if (Objects.isNull(columns)) {
        columns = query.columns;
      } else {
        columns = columns.span(query.columns);
      }
      if (Objects.isNull(rows)) {
        rows = query.rows;
      } else {
        rows = rows.span(query.rows);
      }
      if (Objects.nonNull(gridName) && !gridName.equals(query.getGrid().get())) {
        throw new ValueException("Different grid names.");
      } else {
        gridName = query.getGrid().get();
      }
      return this;
    }

    public CellQuery build() {
      return new CellQuery(gridName, Objects.requireNonNull(columns), Objects.requireNonNull(rows));
    }
  }
}

File: src/main/js/models/nodes/CellQuery.ts
import { ValueException } from "../../errors/ValueException";
import { AlphaUtils } from "../../utils/AlphaUtils";
import { Range } from "../common/Range";
import { isNull } from "../../common/utils/Types";

/**
 * A cell query represents a query of one or more cells over a single sheet. It is composed of
 * two-dimensional ranges: rows and columns, both of which can be unbounded at the lower, or
 * upper ends of the range.
 */
export class CellQuery extends Object {
  /**
   * Range of columns.
   */
  readonly columns: Range;
  /**
   * Range of rows.
   */
  readonly rows: Range;
  /**
   * Name of the sheet being queried.
   */
  private sheet: string = null;

  constructor(sheet: string, columns: Range, rows: Range) {
    super();
    this.sheet = sheet;
    this.columns = columns;
    this.rows = rows;
  }

  static builder(query?: CellQuery): CellQueryBuilder {
    return new CellQueryBuilder(query);
  }

  getFormattedSheetName(): string {
    if (this.sheet && this.sheet.startsWith("'")) {
      return this.sheet.substr(1, this.sheet.length - 2);
    }
    return this.sheet;
  }

  /**
   * Does this query intersect with the other one?
   *
   * @param other - other query.
   * @return true if at least one cell overlaps.
   */
  intersects(other: CellQuery): boolean {
    return other.columns.isConnected(this.columns) && other.rows.isConnected(this.rows);
  }

  /**
   * Bound this query using an upper column and upper row. Will default to the lower of the upper column passed in,
   * and the upper column bound existing on the columns range. Same for rows.
   *
   * @param upperColumn - upper endpoint for columns to default to.
   * @param upperRow    - upper endpoint of rows to default to.
   * @return bounded query.
   */
  toBounded(upperColumn: number, upperRow: number): CellQuery {
    if (
      !Range.closed(0, upperColumn).isConnected(this.columns) &&
      !Range.closed(0, upperRow).isConnected(this.rows)
    ) {
      return this;
    }
    return CellQuery.builder(this)
      .columnsBetween(
        this.columns.lowerEndpoint(),
        Math.min(
          upperColumn,
          this.columns.hasUpperBound() ? this.columns.upperEndpoint() : upperColumn
        )
      )
      .rowsBetween(
        this.rows.lowerEndpoint(),
        Math.min(upperRow, this.rows.hasUpperBound() ? this.rows.upperEndpoint() : upperRow)
      )
      .build();
  }

  toString() {
    const sheet = isNull(this.sheet) ? "" : `${this.sheet}!`;
    const upperColumn = this.columns.hasUpperBound()
      ? AlphaUtils.oneIndexedNumberToLetter(this.columns.upper + 1)
      : "";
    const lowerColumn = this.columns.hasLowerBound()
      ? AlphaUtils.oneIndexedNumberToLetter(this.columns.lower + 1)
      : "";
    const upperRow = this.rows.hasUpperBound() ? (this.rows.upper + 1).toString() : "";
    const lowerRow = this.rows.hasLowerBound() ? (this.rows.lower + 1).toString() : "";
    return `${sheet}${lowerColumn}${lowerRow}:${upperColumn}${upperRow}`;
  }
}

/**
 * Easier way to build a query.
 */
class CellQueryBuilder {
  private sheet: string = null;
  private columns: Range;
  private rows: Range;

  constructor(cellQuery?: CellQuery) {
    if (cellQuery) {
      this.sheet = cellQuery.getFormattedSheetName();
      this.columns = cellQuery.columns;
      this.rows = cellQuery.rows;
    }
  }

  setSheet(sheet: string): CellQueryBuilder {
    this.sheet = sheet;
    return this;
  }

  columnsBetween(low: number, high: number): CellQueryBuilder {
    this.columns = Range.closed(low, high);
    return this;
  }

  columnsBetweenAlpha(low: string, high: string): CellQueryBuilder {
    this.columns = Range.closed(AlphaUtils.columnToInt(low), AlphaUtils.columnToInt(high));
    return this;
  }

  rowsBetween(low: number, high: number): CellQueryBuilder {
    this.rows = Range.closed(low, high);
    return this;
  }

  rowsBetweenAlpha(low: string, high: string): CellQueryBuilder {
    this.rows = Range.closed(parseInt(low) - 1, parseInt(high) - 1);
    return this;
  }

  columnsStartingAt(low: number): CellQueryBuilder {
    this.columns = Range.atLeast(low);
    return this;
  }

  rowsStartingAt(low: number): CellQueryBuilder {
    this.rows = Range.atLeast(low);
    return this;
  }

  openColumnsStartingAtZero(): CellQueryBuilder {
    this.columns = Range.atLeast(0);
    return this;
  }

  openColumnsStartingAt(columnString: string): CellQueryBuilder {
    this.columns = Range.atLeast(AlphaUtils.columnToInt(columnString));
    return this;
  }

  openColumnsStartingAtNumber(column: number): CellQueryBuilder {
    this.columns = Range.atLeast(column);
    return this;
  }

  openRowsStartingAtZero(): CellQueryBuilder {
    this.rows = Range.atLeast(0);
    return this;
  }

  openRowsStartingAtNumber(row: number): CellQueryBuilder {
    this.rows = Range.atLeast(row);
    return this;
  }

  openRowsStartingAt(rowString: string): CellQueryBuilder {
    this.rows = Range.atLeast(AlphaUtils.rowToInt(rowString));
    return this;
  }

  expand(query: CellQuery): CellQueryBuilder {
    if (this.columns == null) {
      this.columns = query.columns;
    } else {
      this.columns = this.columns.span(query.columns);
    }
    if (this.rows == null) {
      this.rows = query.rows;
    } else {
      this.rows = this.rows.span(query.rows);
    }
    if (this.sheet !== null && this.sheet !== query.getFormattedSheetName()) {
      throw new ValueException("Different sheet names.");
    } else {
      this.sheet = query.getFormattedSheetName();
    }

    return this;
  }

  build(): CellQuery {
    return new CellQuery(this.sheet, this.columns, this.rows);
  }
}
